### ISSUE TITLE
State machine executor doesn't recognize new SMs immediately and gets stuck

### ISSUE THREAD

**Issue by elsirion** [fedimint/fedimint#4459]
It appears that the SM executor

1. Doesn't always recognize new SMs immediately while others are running
2. Blocks if too many run at the same time

See the logs attached below.

For 1. I'll rely on @dpc for help since maybe that's benign and expected after the recent refactor. We should think about cases where this could lead to stuck SMs that spawn another SM and then wait for it to finish.

My initial theory for 2 isn't supported by the logs, there should be warnings about autocommit failing.

~For 2. I have an idea what could cause it: since even state transitions run in parallel now there could be enormous amounts or retrying on DB commit failure. In the worst case this roughly scales quadratically with the number of state machines contending for the same resource afaik.~

~Every round the executor runs all remaining state transitions in parallel, so consuming `n` units of CPU time the first round, but only committing 1 of n results to DB. Next round consumes `n-1` CPU, the one after `n-2` and so on.~

![2024-03-05-150446_2880x1800_scrot](https://github.com/fedimint/fedimint/assets/84478054/b77b45c2-c0d2-44de-9f33-037eb7766a7c)

**Comment by dpc** [comment]
@elsirion `Awaiting any state transitions to become ready` is a message from the pre-#4230 codebase. One reason why I rewrote it is because it was prone to problems like this.

After #4230 everything should be much more snappy and smooth. However there is currently no bound on number of futures that can be running in parallel. I figured that under normal circumstances it should stay below 10-thousands-like order of magnitude and I expect tokio to handle it just fine.
